test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+        basic_type_names[i] = (char *) malloc(sizeof(**basic_type_names) * BASIC_TYPE_NAME_MAXLEN);
+        if (!basic_type_names[i]) {
+            err = DTP_ERR_OTHER;
+            fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+            fflush(stdout);
+            goto fn_fail;
+        }
+    }
+
+    /* compute struct size */
+    basic_type_counts = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type_count;
+    basic_types = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type;
+    for (i = 0; i < basic_type_count; i++) {
+        err = MPI_Type_size(basic_types[i], &basic_type_sizes[i]);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        err = MPI_Type_get_name(basic_types[i], basic_type_names[i], &len);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        size += (basic_type_sizes[i] * basic_type_counts[i]);
+    }
+
+    /* compute displacements */
+    displs = 0;
+    for (i = 0; i < basic_type_count; i++) {
+        basic_type_displs[i] = displs;
+        displs += basic_type_sizes[i] * basic_type_counts[i];
+    }
+
+    /* allocate buffer */
+    buf = (void *) malloc(size);
+    if (!buf) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+    memset(buf, 0, size);
+
+    /* create struct datatype */
+    err = MPI_Type_struct(basic_type_count, basic_type_counts, basic_type_displs, basic_types,
+                          &dtp->DTP_obj_array->DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    err = MPI_Type_commit(&dtp->DTP_obj_array->DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    dtp->DTP_obj_array->DTP_obj_count = 1;
+
+    /* initialize every elem in struct separately */
+    for (i = 0; i < basic_type_count; i++) {
+        par->type_count = basic_type_counts[i];
+        par->type_stride = 1;
+        par->type_blklen = 1;
+        par->type_offset = basic_type_displs[i];        /* expressed in bytes */
+        DTPI_Type_init_buf(par, (const char *) basic_type_names[i], buf);
+    }
+
+    dtp->DTP_obj_array->DTP_obj_buf = buf;
+    dtp->DTP_obj_array->private_info = dtpi;
+
+    /* initialize private datatype info data */
+    dtpi->obj_type = DTPI_OBJ_TYPE__STRUCT;
+    dtpi->type_extent = size;
+    dtpi->type_lb = 0;
+    dtpi->type_ub = size;
+    dtpi->u.structure.displs = basic_type_displs;       /* freed in DTP_obj_free() */
+
+    /* set type name for debug information */
+    memset(type_name, 0, sizeof(type_name));
+    sprintf(type_name, "%s (%d elements)", "struct", basic_type_count);
+    err = MPI_Type_set_name(dtp->DTP_obj_array->DTP_obj_type, (char *) type_name);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+
+    for (i = 0; i < basic_type_count; i++) {
+        free(basic_type_names[i]);
+    }
+    if (basic_type_names) {
+        free(basic_type_names);
+    }
+
+  fn_exit:
```

Comment:
- `basic_type_sizes` is not freed.
- Since both `fn_fail` and `fn_exit` free the temporary buffers, why not free them under `fn_exit` thus you can remove the corresponding free calls under `fn_fail`.